### PR TITLE
TL/UCP: use knomial pattern in gather

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Use knomial pattern data structure in gather knomial algorithm in TL UCP

## Why ?
part of https://github.com/openucx/ucc/pull/1035
_

### CODE REVIEW

**Comment by manjugv** [comment]
@aamirshafi  Did you have any comments?

File: src/coll_patterns/sra_knomial.h
Comment (aamirshafi) [inline src/coll_patterns/sra_knomial.h:1]: @Sergei-Lebedev and I discussed this in our review. The filename has `sra_` in it while this code will also be used to implement other collective algorithms (like `srg`). One suggestion will be to relocate code from this file to `src/coll_patterns/recursive_knomial.h` to have a single knomial pattern file for all collectives. Another option can be to have two separate pattern header files for SRG and SRA.

Comment (aamirshafi) [reply src/coll_patterns/sra_knomial.h:1]: Also the functions in `sra_knomial.h` use slightly different function names than `recursive_knomial.h`. Note `ucc_kn_g_pattern_init()` vs. `ucc_knomial_pattern_init`.

File: src/components/tl/ucp/gather/gather_knomial.c [hunk 1]
```diff
@@ -9,135 +9,195 @@
...
-                        status = ucc_mc_memcpy(task->gather_kn.scratch,
-                                            args->src.info.buffer, msg_size,
-                                            args->src.info.mem_type, mtype);
-                    } else if (!UCC_IS_INPLACE(*args)) {
-                        status = ucc_mc_memcpy(
-                            PTR_OFFSET(task->gather_kn.scratch, data_size * rank),
-                            args->src.info.buffer, msg_size,
-                            args->src.info.mem_type, mtype);
-                    }
+            if ((ct != UCC_COLL_TYPE_REDUCE)  &&
```
Comment (aamirshafi) [inline src/components/tl/ucp/gather/gather_knomial.c:131]: Please add a comment why reduce collective is checked for in a gather related file.

**Review by aamirshafi — CHANGES_REQUESTED** [review]
@Sergei-Lebedev presented this PR in couple of sessions. I have requested a few changes that are inserted as comments: 1) clarify the structure of header files ([recursive_knomial.h](https://github.com/Sergei-Lebedev/ucc/blob/topic/gather_kn_pattern/src/coll_patterns/recursive_knomial.h) and [sra_knomial.h](https://github.com/Sergei-Lebedev/ucc/blob/topic/gather_kn_pattern/src/coll_patterns/sra_knomial.h)), and 2) add a few comments to explain corner-cases.

**Comment by aamirshafi** [comment]
> @aamirshafi Did you have any comments?

I have added my review @manjugv.

**Review by samnordmann — APPROVED** [review]

**Comment by aamirshafi** [comment]
@manjugv - this PR looks good to me.

**Review by janjust — APPROVED** [review]

**Pull request merged**